src/databricks/labs/ucx/hive_metastore/locations.py

```diff
@@ -154,17 +154,60 @@ def __init__(
         schema: str,
         tables_crawler: TablesCrawler,
         mounts_crawler: 'MountsCrawler',
+        enable_hms_federation: bool = False,
     ):
         super().__init__(sql_backend, "hive_metastore", schema, "external_locations", ExternalLocation)
         self._ws = ws
         self._tables_crawler = tables_crawler
         self._mounts_crawler = mounts_crawler
+        self._enable_hms_federation = enable_hms_federation
 
     @cached_property
     def _mounts_snapshot(self) -> list['Mount']:
         """Returns all mounts, sorted by longest prefixes first."""
         return sorted(self._mounts_crawler.snapshot(), key=lambda _: (len(_.name), _.name), reverse=True)
 
+    @staticmethod
+    def clean_location(location: str) -> str:
+        # remove the s3a scheme and replace it with s3 as these can be considered the same and will be treated as such
+        # Having s3a and s3 as separate locations will cause issues when trying to find overlapping locations
+        return re.sub(r"^s3a:/", r"s3:/", location).rstrip("/")
+
+    def external_locations_with_root(self) -> list[ExternalLocation]:
+        # Returns a list of external locations with the DBFS root location appended to the list
+        # Used for HMS Federation use cases
+        external_locations = list(self.snapshot())
+        if not self._enable_hms_federation:
+            return external_locations
+        dbfs_root = self._get_dbfs_root()
+        if dbfs_root:
+            external_locations.append(dbfs_root)
+        return external_locations
+
+    def _get_dbfs_root(self) -> ExternalLocation | None:
+        """
+        Get the root location of the DBFS only if HMS Fed is enabled.
+        Utilizes an undocumented Databricks API call
+
+        Returns:
+            Cloud storage root location for dbfs
+
+        """
+        if not self._enable_hms_federation:
+            return None
+        logger.debug("Retrieving DBFS root location")
+        try:
+            response = self._ws.api_client.do("GET", "/api/2.0/dbfs/resolve-path", query={"path": "dbfs:/"})
+            if isinstance(response, dict):
+                resolved_path = response.get("resolved_path")
+                if resolved_path:
+                    return ExternalLocation(self.clean_location(resolved_path), 0)
+        except NotFound:
```

Comment:
Did you test if this is the expected exception?

Reply:
Yes